### PR TITLE
Fix for a possible memory leak in observer

### PR DESCRIPTION
**Pull request by hzzcc**
Reaction cause memory leak

### Code change checklist

-   [ ] Added/updated unit tests
-   [ ] Updated `README` if applicable

### CODE REVIEW

**Comment by FredyC** [comment]
Please fix error in failing build. And also can you elaborate on why this change? Did you encounter some issue that you are able to reproduce? In that case, can you add a test that covers that issue?

**Comment by mweststrate** [comment]
Thanks for taking the effort of creating this PR, but please elaborate on how this prevents a memory leak, as we can't accept changes without some rationally or verification

**Comment by hzzcc** [comment]
image
image

**Comment by FredyC** [comment]
Sorry, but one screenshot and `tsignore` (😮) in the code won't do. Can you explain the problem in words?

File: src/useObserver.ts [hunk 1]
```diff
@@ -67,7 +67,9 @@ export function useObserver<T>(
                 // We haven't yet reached useEffect(), so we'll need to trigger a re-render
                 // when (and if) useEffect() arrives.  The easiest way to do that is just to
                 // drop our current reaction and allow useEffect() to recreate it.
-                newReaction.dispose()
+                reactionTrackingRef.current!.reaction.dispose()
+                // @ts-ignore
+                reactionTrackingRef.current!.reaction = null
```
Comment (FredyC) [inline src/useObserver.ts:72]: This just doesn't feel like it's necessary. I mean it won't probably harm anything, but I would rather keep code that makes sense only. It's obvious, that calling `dispose` won't remove reaction from the memory immediately. Even removing its reference doesn't give such assurance, it's up to GC.

File: src/useObserver.ts [hunk 2]
```diff
@@ -67,7 +67,9 @@ export function useObserver<T>(
                 // We haven't yet reached useEffect(), so we'll need to trigger a re-render
                 // when (and if) useEffect() arrives.  The easiest way to do that is just to
                 // drop our current reaction and allow useEffect() to recreate it.
-                newReaction.dispose()
```
Comment (FredyC) [inline src/useObserver.ts:70]: Explain why this is harmful to have here, please.

Comment (hzzcc) [reply src/useObserver.ts:70]: I found that when keep "newReaction" in the func，the reference newReaction will still in the GC.
But when delete the newReaction, use reactionTrackingRef.current, the reference will be removed.

Comment (hzzcc) [reply src/useObserver.ts:70]: [https://github.com/hzzcc/mobx-react-lite-test](https://github.com/hzzcc/mobx-react-lite-test)
Here is the test repo.

**Comment by hzzcc** [comment]
[https://github.com/hzzcc/mobx-react-lite-test](https://github.com/hzzcc/mobx-react-lite-test)
Here is the test repo.

**Comment by hzzcc** [comment]
The test flow: 
1. click "onChange"
2. change the input
1. click "onChange" to hide input
3. profile the memory

**Comment by FredyC** [comment]
Sorry, but the test repo is not gonna be enough. If you have discovered a flaw, we need to figure out to reproduce it in a test file and that needs to be added as a part of this PR. The reason is to avoid regressions for future changes.

File: src/useObserver.ts (hunk 1 above)
Comment (FredyC) [reply src/useObserver.ts:72]: @hzzcc Can you respond to this? Not sure why you marked it as resolved without an answer.

Comment (hzzcc) [reply src/useObserver.ts:72]: @FredyC Yes, this is up to GC, but  can take heap snapshot by youself, click several times until memory stabilizes.

Comment (FredyC) [reply src/useObserver.ts:72]: I don't follow. How does that explain why do you need to set a reaction to null here?

Comment (hzzcc) [reply src/useObserver.ts:72]: Dereference 'reaction'. But, I guess input change event cause these. when I change to 'div', the problem disappear.

Comment (FredyC) [reply src/useObserver.ts:72]: Ok, so please remove this from PR if it's not necessary, it only brings confusion.

**Comment by FredyC** [comment]
I will keep it open as it might worth investigating if there is an actual memory leak. Help is certainly wanted here.

**Comment by FredyC** [comment]
We now have an alternative solution #332 so it's probably not worth it to worry about a memory leak in here. Besides nobody else seem to have noticed anything.

**Pull request closed without merging**